File: scripts/detect_fuel_anomalies.py
#!/usr/bin/env python3
"""Fuel Price Anomaly Detector — flags overpay fuel transactions >15% above fleet average.

Read-only. Queries the bcat-ops AppSync GraphQL API for fuel transactions (last 30 days),
computes fleet average price/gallon per fuel type, flags transactions exceeding the
threshold, and outputs flagged transactions as JSON.

Authentication: Cognito USER_PASSWORD_AUTH using BCAT_EMAIL / BCAT_PASSWORD env vars
(same auth as iv1 ingestFuelReport.mjs).

Usage:
    BCAT_EMAIL=ryne@bcatcorp.com BCAT_PASSWORD=... python3 detect_fuel_anomalies.py
    BCAT_EMAIL=... BCAT_PASSWORD=... python3 detect_fuel_anomalies.py --threshold 15
    BCAT_EMAIL=... BCAT_PASSWORD=... python3 detect_fuel_anomalies.py --days 30 --threshold 15 --json
"""

import json
import os
import sys
import datetime
import argparse
from typing import Optional

import boto3
from botocore.exceptions import ClientError

# ── Configuration ────────────────────────────────────────────────────────────────
APPSYNC_URL = "https://2ivs344vbvglblxbmjdd7pgi2q.appsync-api.us-east-1.amazonaws.com/graphql"
USER_POOL_ID = "us-east-1_IbPKPNJC9"
CLIENT_ID = "1qr651dshigrv8a6c338d4f709"
AWS_REGION = "us-east-1"

# Fuel types recognized as real fuel (matches bcat-ops isFuelTx / FUEL_ITEM_TYPES)
FUEL_ITEM_TYPES = {"ULSD", "FUEL", "DEFD", "BIO", "B5", "B20", "REG", "PREM", "DSL"}

# ── GraphQL queries ──────────────────────────────────────────────────────────────
LIST_FUEL_QUERY = """
query ListFuelTransactions($startDate: String, $endDate: String) {
  listFuelTransactions(limit: 10000, filter: {
    transactionDate: { between: [$startDate, $endDate] }
  }) {
    items {
      id transactionDate cardNumber invoiceNumber unitNumber truckId driverName
      odometer locationName city state fees fuelType itemCategory pricePerUnit quantity amount
      currency sourceFile importedAt createdAt updatedAt
    }
  }
}
"""

# Fallback: if AppSync filter isn't supported, list everything and filter client-side.
LIST_ALL_QUERY = """
query ListFuelTransactions {
  listFuelTransactions(limit: 10000) {
    items {
      id transactionDate cardNumber invoiceNumber unitNumber truckId driverName
      odometer locationName city state fees fuelType itemCategory pricePerUnit quantity amount
      currency sourceFile importedAt createdAt updatedAt
    }
  }
}
"""


def authenticate(email: str, password: str) -> str:
    """Sign in to Cognito and return the ID token."""
    client = boto3.client("cognito-idp", region_name=AWS_REGION)

    try:
        resp = client.initiate_auth(
            AuthFlow="USER_PASSWORD_AUTH",
            AuthParameters={"USERNAME": email, "PASSWORD": password},
            ClientId=CLIENT_ID,
        )
    except ClientError as e:
        code = e.response["Error"]["Code"]
        if code == "NotAuthorizedException":
            raise SystemExit(f"Authentication failed: invalid credentials ({e})") from e
        if code == "UserNotFoundException":
            raise SystemExit(f"Authentication failed: user not found ({e})") from e
        raise

    id_token = resp["AuthenticationResult"]["IdToken"]
    return id_token


def graphql_call(query: str, variables: dict, id_token: str) -> dict:
    """Make a GraphQL request to AppSync with Cognito auth."""
    import urllib.request

    body = json.dumps({"query": query, "variables": variables}).encode("utf-8")
    req = urllib.request.Request(
        APPSYNC_URL,
        data=body,
        headers={
            "Content-Type": "application/json",
            "Authorization": id_token,
        },
    )
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read())


def is_fuel(tx: dict) -> bool:
    """Check if a transaction is genuine fuel (mirrors driverFuel.isFuelTx)."""
    cat = (tx.get("itemCategory") or "").strip()
    if cat:
        return cat == "FUEL"
    return (tx.get("fuelType") or "").upper().strip() in FUEL_ITEM_TYPES


def fetch_fuel_transactions(id_token: str, days: int) -> list[dict]:
    """Fetch fuel transactions from the last N days."""
    end_date = datetime.date.today()
    start_date = end_date - datetime.timedelta(days=days)
    start_iso = start_date.isoformat()
    end_iso = end_date.isoformat()

    # Try with date filter first; fall back to unfiltered if AppSync rejects the filter
    for attempt, (query, variables) in enumerate(
        [
            (LIST_FUEL_QUERY, {"startDate": start_iso, "endDate": end_iso}),
            (LIST_ALL_QUERY, {}),
        ],
        1,
    ):
        try:
            result = graphql_call(query, variables, id_token)
        except Exception as e:
            if attempt < 2:
                print(f"[warn] Filtered query failed ({e}), falling back to unfiltered...", file=sys.stderr)
                continue
            raise

        if result.get("errors"):
            msg = result["errors"][0].get("message", str(result["errors"]))
            if attempt < 2 and ("filter" in msg.lower() or "between" in msg.lower()):
                print(f"[warn] Filter not supported by backend, retrying unfiltered...", file=sys.stderr)
                continue
            raise SystemExit(f"GraphQL error: {msg}")

        items = result.get("data", {}).get("listFuelTransactions", {}).get("items", []) or []
        break
    else:
        # Shouldn't happen (both attempts either break or raise), but satisfy the type checker
        raise SystemExit("Failed to fetch fuel transactions: unexpected error")

    # If unfiltered query was used, filter client-side
    if days:
        items = [tx for tx in items if tx.get("transactionDate", "") >= start_iso and tx.get("transactionDate", "") <= end_iso]

    # Filter to real fuel only
    fuel_txs = [tx for tx in items if is_fuel(tx)]

    return fuel_txs


def dedup_transactions(txs: list[dict]) -> list[dict]:
    """De-duplicate using invoice-agnostic key (date|card|fuelType|amount|quantity)."""
    seen = set()
    deduped = []
    for tx in txs:
        key = (
            f"{tx.get('transactionDate', '')}|"
            f"{tx.get('cardNumber', '')}|"
            f"{tx.get('fuelType', '')}|"
            f"{tx.get('amount', 0)}|"
            f"{tx.get('quantity', 0)}"
        )
        if key not in seen:
            seen.add(key)
            deduped.append(tx)
    return deduped


def normalize_card(card: Optional[str]) -> str:
    """Digits only, leading zeros stripped for card matching."""
    import re
    return re.sub(r"\D", "", (card or "")).lstrip("0")


def detect_anomalies(txs: list[dict], threshold_pct: float) -> dict:
    """Compute fleet averages and flag overpay transactions.

    Returns: {
        "fleet_averages": {fuelType: {"avg_price_per_unit": float, "count": int, "total_gallons": float}},
        "flags": [flag, ...],
        "summary": {...}
    }
    """
    # Group by fuel type for averaging
    by_fuel_type: dict[str, list[dict]] = {}
    for tx in txs:
        ft = (tx.get("fuelType") or "UNKNOWN").upper().strip()
        by_fuel_type.setdefault(ft, []).append(tx)

    fleet_averages = {}
    for ft, group in sorted(by_fuel_type.items()):
        prices = [tx["pricePerUnit"] for tx in group if tx.get("pricePerUnit", 0) > 0]
        total_gallons = sum(tx.get("quantity", 0) for tx in group)
        if prices:
            avg = sum(prices) / len(prices)
        else:
            avg = 0.0
        fleet_averages[ft] = {
            "avg_price_per_unit": round(avg, 4),
            "count": len(group),
            "total_gallons": round(total_gallons, 2),
        }

    flags = []
    for tx in txs:
        ft = (tx.get("fuelType") or "UNKNOWN").upper().strip()
        ppu = tx.get("pricePerUnit", 0)
        avg_info = fleet_averages.get(ft, {})
        fleet_avg = avg_info.get("avg_price_per_unit", 0)

        if fleet_avg <= 0 or ppu <= 0:
            continue

        overpay_pct = ((ppu - fleet_avg) / fleet_avg) * 100
        if overpay_pct > threshold_pct:
            overpay_amount = round((ppu - fleet_avg) * tx.get("quantity", 0), 2)
            flags.append({
                "id": tx.get("id"),
                "transaction_date": tx.get("transactionDate"),
                "truck_unit": tx.get("unitNumber") or (f"TRK-{normalize_card(tx.get('cardNumber'))}" if tx.get("cardNumber") else "Unknown"),
                "truck_id": tx.get("truckId"),
                "driver": tx.get("driverName"),
                "location": f"{tx.get('city', '')} {tx.get('state', '')}".strip(),
                "fuel_type": ft,
                "price_per_gallon": ppu,
                "fleet_average": round(fleet_avg, 4),
                "overpay_pct": round(overpay_pct, 2),
                "overpay_amount": overpay_amount,
                "gallons": tx.get("quantity", 0),
                "total_cost": tx.get("amount", 0),
            })

    # Sort by overpay percentage descending
    flags.sort(key=lambda f: f["overpay_pct"], reverse=True)

    return {
        "fleet_averages": fleet_averages,
        "flags": flags,
        "summary": {
            "total_transactions": len(txs),
            "total_flagged": len(flags),
            "threshold_pct": threshold_pct,
            "date_range_days": (datetime.date.today() - datetime.date.fromisoformat(min(tx.get("transactionDate", "2000-01-01")[:10] for tx in txs))).days
                if txs else 0,
        },
    }


def format_slack_message(results: dict) -> str:
    """Format flagged transactions as a Slack message."""
    flags = results["flags"]
    summary = results["summary"]
    averages = results["fleet_averages"]

    if not flags:
        return (
            f":fuelpump: *Fuel Anomaly Report* — {datetime.date.today().isoformat()}\n"
            f"✅ No overpay transactions detected (>15% above fleet average).\n"
            f"_{summary['total_transactions']} transactions analyzed across {len(averages)} fuel types._"
        )

    lines = [
        f":warning: *Fuel Anomaly Report* — {datetime.date.today().isoformat()}",
        f"*{len(flags)} overpay transaction(s)* detected (>15% above fleet average) out of {summary['total_transactions']} analyzed.",
        "",
    ]

    for flag in flags:
        loc = flag["location"] or "Unknown location"
        unit = flag["truck_unit"] or "Unknown unit"
        lines.append(
            f"• *{unit}*: ${flag['price_per_gallon']:.2f}/gal in {loc} "
            f"— {flag['overpay_pct']:.0f}% above fleet avg (${flag['fleet_average']:.2f}/gal). "
            f"Overpay: ${flag['overpay_amount']:.2f}"
        )

    lines.append("")
    lines.append("*Fleet Averages (30-day):*")
    for ft, info in sorted(averages.items()):
        lines.append(f"  • {ft}: ${info['avg_price_per_unit']:.2f}/gal ({info['count']} txns, {info['total_gallons']:.1f} gal)")

    return "\n".join(lines)


def main():
    parser = argparse.ArgumentParser(description="Fuel price anomaly detector for bcat-ops")
    parser.add_argument("--threshold", type=float, default=15.0, help="Overpay threshold percent (default: 15)")
    parser.add_argument("--days", type=int, default=30, help="Lookback window in days (default: 30)")
    parser.add_argument("--json", action="store_true", help="Output raw JSON instead of formatted text")
    parser.add_argument("--slack", action="store_true", help="Output Slack-formatted markdown message")
    args = parser.parse_args()

    email = os.environ.get("BCAT_EMAIL")
    password = os.environ.get("BCAT_PASSWORD")
    if not email or not password:
        print("Error: BCAT_EMAIL and BCAT_PASSWORD environment variables are required.", file=sys.stderr)
        print("       Source them from ~/.hermes/bcat-ops-credentials.env", file=sys.stderr)
        sys.exit(1)

    print(f"Authenticating as {email}...", file=sys.stderr)
    id_token = authenticate(email, password)

    print(f"Fetching fuel transactions (last {args.days} days)...", file=sys.stderr)
    txs = fetch_fuel_transactions(id_token, args.days)
    print(f"  Fetched {len(txs)} fuel transactions.", file=sys.stderr)

    deduped = dedup_transactions(txs)
    if len(deduped) < len(txs):
        print(f"  De-duplicated: {len(txs)} → {len(deduped)} unique.", file=sys.stderr)

    if not deduped:
        print("No fuel transactions found in the lookback window.", file=sys.stderr)
        if args.json:
            print(json.dumps({"fleet_averages": {}, "flags": [], "summary": {"total_transactions": 0, "total_flagged": 0}}))
        else:
            print("No fuel transactions found in the lookback window.")
        return

    results = detect_anomalies(deduped, args.threshold)

    if args.json:
        print(json.dumps(results, indent=2, default=str))
    elif args.slack:
        print(format_slack_message(results))
    else:
        # Default: human-readable output
        summary = results["summary"]
        print(f"\nFuel Anomaly Detection Report")
        print(f"=============================")
        print(f"Period: last {args.days} days ({summary['date_range_days']} days of data)")
        print(f"Transactions analyzed: {summary['total_transactions']}")
        print(f"Threshold: {summary['threshold_pct']}% above fleet average")
        print(f"Flagged: {summary['total_flagged']}")
        print()

        print("Fleet Averages (price/gallon):")
        for ft, info in sorted(results["fleet_averages"].items()):
            print(f"  {ft:6s}: ${info['avg_price_per_unit']:.4f}/gal  ({info['count']} txns, {info['total_gallons']:.1f} gal)")

        if results["flags"]:
            print(f"\n⚠️  Flagged Overpay Transactions (>{(args.threshold)}% above fleet average):")
            for flag in results["flags"]:
                print(f"  {flag['truck_unit']:12s} | {flag['transaction_date']} | {flag['location']:30s} | "
                      f"${flag['price_per_gallon']:.2f}/gal | "
                      f"+{flag['overpay_pct']:.1f}% | overpay: ${flag['overpay_amount']:.2f} | "
                      f"{flag['gallons']:.1f} gal | ${flag['total_cost']:.2f} total")
        else:
            print(f"\n✅ No overpay transactions detected.")


if __name__ == "__main__":
    main()

File: scripts/detectFuelAnomalies.mjs
#!/usr/bin/env node
/**
 * Fuel Price Anomaly Detector — flags overpay fuel transactions >15% above fleet average.
 *
 * READ-ONLY. Queries the bcat-ops AppSync GraphQL API for fuel transactions (last 30 days),
 * computes fleet average price/gallon per fuel type, and flags any transaction where
 * pricePerUnit exceeds 15% above the fleet average.
 *
 * Auth: Cognito signIn via aws-amplify (same auth as iv1 ingestFuelReport.mjs).
 *
 * Usage:
 *   source ~/.hermes/bcat-ops-credentials.env
 *   node scripts/detectFuelAnomalies.mjs [--threshold 15] [--days 30] [--json] [--slack] [--post-to-slack]
 *
 * Flags:
 *   --threshold N    Overpay threshold percent (default: 15)
 *   --days N          Lookback window in days (default: 30)
 *   --json            Output raw JSON instead of formatted text
 *   --slack           Output Slack-formatted markdown message to stdout
 *   --post-to-slack   POST the Slack message via SLACK_WEBHOOK_URL env var
 *
 * For cron: wrapper at ~/.hermes/scripts/fuel-anomaly-check.sh (cron id: aa33291bf992)
 */

import { readFileSync } from 'fs'
import { dirname, resolve } from 'path'
import { fileURLToPath } from 'url'
import { Amplify } from 'aws-amplify'
import { signIn, fetchAuthSession } from 'aws-amplify/auth'

const __dirname = dirname(fileURLToPath(import.meta.url))

// ── Amplify config ─────────────────────────────────────────────────────────────
const outputs = JSON.parse(readFileSync(resolve(__dirname, '../amplify_outputs.json'), 'utf8'))
const APPSYNC_URL = outputs.data.url
Amplify.configure(outputs)

// Fuel types recognized as real fuel (matches driverFuel.ts FUEL_ITEM_TYPES)
const FUEL_ITEM_TYPES = new Set(['ULSD', 'FUEL', 'DEFD', 'BIO', 'B5', 'B20', 'REG', 'PREM', 'DSL'])

// ── GraphQL ────────────────────────────────────────────────────────────────────
const LIST_FUEL_QUERY = `query ListFuelTransactions($filter: ModelFuelTransactionFilterInput) {
  listFuelTransactions(limit: 10000, filter: $filter) {
    items {
      id transactionDate cardNumber invoiceNumber unitNumber truckId driverName
      odometer locationName city state fees fuelType itemCategory pricePerUnit quantity amount
      currency sourceFile importedAt createdAt updatedAt
    }
  }
}`

const LIST_ALL_QUERY = `query ListFuelTransactions {
  listFuelTransactions(limit: 10000) {
    items {
      id transactionDate cardNumber invoiceNumber unitNumber truckId driverName
      odometer locationName city state fees fuelType itemCategory pricePerUnit quantity amount
      currency sourceFile importedAt createdAt updatedAt
    }
  }
}`

async function callAppSync(query, variables, idToken) {
  const res = await fetch(APPSYNC_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: idToken },
    body: JSON.stringify({ query, variables }),
  })
  return res.json()
}

// ── Fuel detection (mirrors driverFuel.ts) ─────────────────────────────────────
function isFuelTx(tx) {
  const cat = (tx.itemCategory || '').trim()
  if (cat) return cat === 'FUEL'
  return FUEL_ITEM_TYPES.has((tx.fuelType || '').toUpperCase().trim())
}

function normalizeCard(card) {
  return (card || '').replace(/\D/g, '').replace(/^0+/, '')
}

// ── Main logic ─────────────────────────────────────────────────────────────────
async function fetchFuelTransactions(idToken, days) {
  const endDate = new Date()
  const startDate = new Date(endDate - days * 86400000)
  const startIso = startDate.toISOString().slice(0, 10)
  const endIso = endDate.toISOString().slice(0, 10)

  // Try with date filter first; fall back to unfiltered
  for (const [query, variables] of [
    [LIST_FUEL_QUERY, { filter: { transactionDate: { between: [startIso, endIso] } } }],
    [LIST_ALL_QUERY, {}],
  ]) {
    try {
      const result = await callAppSync(query, variables, idToken)
      if (result.errors) {
        const msg = result.errors[0].message
        if (variables.filter && (msg.includes('filter') || msg.includes('between'))) {
          console.warn('[warn] Filter not supported by backend, retrying unfiltered...')
          continue
        }
        throw new Error(`GraphQL error: ${msg}`)
      }
      let items = result.data.listFuelTransactions.items || []

      // If unfiltered, filter client-side
      if (!variables.filter && days) {
        items = items.filter(tx => tx.transactionDate >= startIso && tx.transactionDate <= endIso)
      }

      // Filter to fuel only
      return items.filter(isFuelTx)
    } catch (err) {
      if (variables.filter) {
        console.warn(`[warn] Filtered query failed (${err.message}), retrying unfiltered...`)
        continue
      }
      throw err
    }
  }
  throw new Error('Failed to fetch fuel transactions')
}

function dedupTransactions(txs) {
  const seen = new Set()
  return txs.filter(tx => {
    const key = `${tx.transactionDate}|${tx.cardNumber}|${tx.fuelType}|${tx.amount}|${tx.quantity}`
    if (seen.has(key)) return false
    seen.add(key)
    return true
  })
}

function detectAnomalies(txs, thresholdPct) {
  // Group by fuel type
  const byType = {}
  for (const tx of txs) {
    const ft = (tx.fuelType || 'UNKNOWN').toUpperCase().trim()
    if (!byType[ft]) byType[ft] = []
    byType[ft].push(tx)
  }

  // Fleet averages
  const fleetAverages = {}
  for (const [ft, group] of Object.entries(byType)) {
    const prices = group.map(tx => tx.pricePerUnit).filter(p => p > 0)
    const avg = prices.length > 0 ? prices.reduce((a, b) => a + b, 0) / prices.length : 0
    const totalGallons = group.reduce((s, tx) => s + (tx.quantity || 0), 0)
    fleetAverages[ft] = {
      avg_price_per_unit: Math.round(avg * 10000) / 10000,
      count: group.length,
      total_gallons: Math.round(totalGallons * 100) / 100,
    }
  }

  // Flag overpay transactions
  const flags = []
  for (const tx of txs) {
    const ft = (tx.fuelType || 'UNKNOWN').toUpperCase().trim()
    const ppu = tx.pricePerUnit || 0
    const fleetAvg = fleetAverages[ft]?.avg_price_per_unit || 0

    if (fleetAvg <= 0 || ppu <= 0) continue

    const overpayPct = ((ppu - fleetAvg) / fleetAvg) * 100
    if (overpayPct > thresholdPct) {
      const overpayAmount = Math.round((ppu - fleetAvg) * (tx.quantity || 0) * 100) / 100
      flags.push({
        id: tx.id,
        transaction_date: tx.transactionDate,
        truck_unit: tx.unitNumber || (tx.cardNumber ? `TRK-${normalizeCard(tx.cardNumber)}` : 'Unknown'),
        truck_id: tx.truckId,
        driver: tx.driverName || '',
        location: [tx.city, tx.state].filter(Boolean).join(' '),
        fuel_type: ft,
        price_per_gallon: ppu,
        fleet_average: Math.round(fleetAvg * 10000) / 10000,
        overpay_pct: Math.round(overpayPct * 100) / 100,
        overpay_amount: overpayAmount,
        gallons: tx.quantity || 0,
        total_cost: tx.amount || 0,
      })
    }
  }

  flags.sort((a, b) => b.overpay_pct - a.overpay_pct)

  return {
    fleet_averages: fleetAverages,
    flags,
    summary: {
      total_transactions: txs.length,
      total_flagged: flags.length,
      threshold_pct: thresholdPct,
    },
  }
}

function formatSlackMessage(results) {
  const { flags, summary, fleet_averages } = results
  const today = new Date().toISOString().slice(0, 10)

  if (!flags.length) {
    return `:fuelpump: *Fuel Anomaly Report* — ${today}\n` +
      `✅ No overpay transactions detected (>15% above fleet average).\n` +
      `_${summary.total_transactions} transactions analyzed across ${Object.keys(fleet_averages).length} fuel types._`
  }

  const lines = [
    `:warning: *Fuel Anomaly Report* — ${today}`,
    `*${flags.length} overpay transaction(s)* detected (>15% above fleet average) out of ${summary.total_transactions} analyzed.\n`,
  ]

  for (const f of flags) {
    lines.push(
      `• *${f.truck_unit}*: $${f.price_per_gallon.toFixed(2)}/gal in ${f.location || 'Unknown location'}` +
      ` — ${Math.round(f.overpay_pct)}% above fleet avg ($${f.fleet_average.toFixed(2)}/gal).` +
      ` Overpay: $${f.overpay_amount.toFixed(2)}`
    )
  }

  lines.push('\n*Fleet Averages (30-day):*')
  for (const [ft, info] of Object.entries(fleet_averages).sort()) {
    lines.push(`  • ${ft}: $${info.avg_price_per_unit.toFixed(2)}/gal (${info.count} txns, ${info.total_gallons.toFixed(1)} gal)`)
  }

  return lines.join('\n')
}

// ── Slack posting (optional) ───────────────────────────────────────────────────
async function postToSlack(text) {
  const webhookUrl = process.env.SLACK_WEBHOOK_URL
  if (!webhookUrl) {
    console.error('[warn] SLACK_WEBHOOK_URL not set — cannot post to Slack.')
    console.error('       Set it to a Slack Incoming Webhook URL to enable auto-posting.')
    return { ok: false, error: 'No webhook URL configured' }
  }
  try {
    const res = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text }),
    })
    if (!res.ok) {
      console.error(`[warn] Slack webhook returned ${res.status} ${res.statusText}`)
      return { ok: false, error: `HTTP ${res.status}` }
    }
    console.error('[ok] Posted to Slack.')
    return { ok: true }
  } catch (err) {
    console.error(`[warn] Slack webhook failed: ${err.message}`)
    return { ok: false, error: err.message }
  }
}

// ── CLI ────────────────────────────────────────────────────────────────────────
function parseArgs(argv) {
  const args = {}
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i]
    if (a === '--threshold' && argv[i + 1]) { args.threshold = parseFloat(argv[++i]) }
    else if (a === '--days' && argv[i + 1]) { args.days = parseInt(argv[++i], 10) }
    else if (a === '--json') { args.json = true }
    else if (a === '--slack') { args.slack = true }
    else if (a === '--post-to-slack') { args.postToSlack = true }
    else if (!a.startsWith('--') && !args.threshold) { args.threshold = parseFloat(a) } // positional threshold (backwards-compat)
  }
  return args
}

async function main() {
  const cli = parseArgs(process.argv.slice(2))
  const threshold = cli.threshold || 15
  const days = cli.days || 30
  const jsonOut = cli.json || false
  const slackOut = cli.slack || false
  const doPost = cli.postToSlack || false

  const email = process.env.BCAT_EMAIL
  const password = process.env.BCAT_PASSWORD
  if (!email || !password) {
    console.error('Error: BCAT_EMAIL and BCAT_PASSWORD environment variables are required.')
    console.error('       Source them from ~/.hermes/bcat-ops-credentials.env')
    process.exit(1)
  }

  console.error(`Authenticating as ${email}...`)
  await signIn({ username: email, password })
  const session = await fetchAuthSession()
  const idToken = session.tokens?.idToken?.toString()
  if (!idToken) {
    console.error('No ID token after sign-in')
    process.exit(1)
  }

  console.error(`Fetching fuel transactions (last ${days} days)...`)
  const txs = await fetchFuelTransactions(idToken, days)
  console.error(`  Fetched ${txs.length} fuel transactions.`)

  const deduped = dedupTransactions(txs)
  if (deduped.length < txs.length) {
    console.error(`  De-duplicated: ${txs.length} → ${deduped.length} unique.`)
  }

  if (!deduped.length) {
    console.error('No fuel transactions found in the lookback window.')
    if (jsonOut) console.log(JSON.stringify({ fleet_averages: {}, flags: [], summary: { total_transactions: 0, total_flagged: 0 } }))
    else console.log('No fuel transactions found in the lookback window.')
    return
  }

  const results = detectAnomalies(deduped, threshold)

  if (jsonOut) {
    console.log(JSON.stringify(results, null, 2))
  } else if (slackOut || doPost) {
    const slackMsg = formatSlackMessage(results)
    if (slackOut) console.log(slackMsg)
    if (doPost) await postToSlack(slackMsg)
  } else {
    // Human-readable
    console.log(`\nFuel Anomaly Detection Report`)
    console.log(`=============================`)
    console.log(`Period: last ${days} days`)
    console.log(`Transactions analyzed: ${results.summary.total_transactions}`)
    console.log(`Threshold: ${threshold}% above fleet average`)
    console.log(`Flagged: ${results.summary.total_flagged}`)
    console.log()

    console.log('Fleet Averages (price/gallon):')
    for (const [ft, info] of Object.entries(results.fleet_averages).sort()) {
      console.log(`  ${ft.padEnd(6)}: $${info.avg_price_per_unit.toFixed(4)}/gal  (${info.count} txns, ${info.total_gallons.toFixed(1)} gal)`)
    }

    if (results.flags.length) {
      console.log(`\n⚠️  Flagged Overpay Transactions (>${threshold}% above fleet average):`)
      for (const f of results.flags) {
        console.log(
          `  ${(f.truck_unit || '').padEnd(12)} | ${f.transaction_date} | ` +
          `${(f.location || '').padEnd(30)} | $${f.price_per_gallon.toFixed(2)}/gal | ` +
          `+${f.overpay_pct.toFixed(1)}% | overpay: $${f.overpay_amount.toFixed(2)} | ` +
          `${f.gallons.toFixed(1)} gal | $${f.total_cost.toFixed(2)} total`
        )
      }
    } else {
      console.log(`\n✅ No overpay transactions detected.`)
    }
  }
}

main().catch(err => {
  console.error(err)
  process.exit(1)
})